src/uri/IpAddress.cpp

```diff
@@ -19,57 +19,145 @@
+ * SPDX-FileCopyrightText: 2023-2024 General Motors GTO LLC
  * SPDX-License-Identifier: Apache-2.0
  */
 
 #include <array>
 #include <arpa/inet.h>
 #include <spdlog/spdlog.h>
-#include <up-cpp/uri/serializer/IpAddress.h>
+#include <up-cpp/uri/tools/IpAddress.h>
+#include "up-core-api/uri.pb.h"
 
-using namespace uprotocol::uri;
+namespace {
+    namespace uri {
+        using namespace uprotocol::uri;
+    }
+
+    using UAuthority = uprotocol::v1::UAuthority;
+
+    /// Controls whether buffers created when converting address formats have
+    /// `shrink_to_fit()` called on them. This should be set to true only if
+    /// it seems reasonably likely that large numbers of IpAddress instances
+    /// will be kept around for long times.
+    constexpr bool OPTION_SHRINK_BUFFERS = false;
+
+    std::vector<uint8_t> bytesFromAuthority(UAuthority const& authority) {
+        std::vector<uint8_t> addressBytes;
+        if (authority.has_ip() && authority.ip().size() > 0) {
+            auto const& addrByteStr = authority.ip();
+            std::transform(addrByteStr.begin(), addrByteStr.end(),
+                    std::back_inserter(addressBytes),
+                    [](char c){ return *reinterpret_cast<uint8_t*>(&c); });
+        }
+        return addressBytes;
+    }
+
+    uri::IpAddress::Type typeFromAuthority(UAuthority const& authority) {
+        if (authority.has_ip()) {
+            if (authority.ip().size() == uri::IpAddress::IpV4AddressBytes) {
+                return uri::IpAddress::Type::IpV4;
+            } else if (authority.ip().size() == uri::IpAddress::IpV6AddressBytes) {
+                return uri::IpAddress::Type::IpV6;
+            }
+            spdlog::error("UAuthority has IP address, but size ({}) does not "
+                    "match expected for IPv4 or IPv6", authority.ip().size());
+        } else {
+            spdlog::error("UAuthority does not have IP address");
+        }
+        return uri::IpAddress::Type::Invalid;
+    }
+}
+
+uri::IpAddress::IpAddress(uprotocol::v1::UAuthority const& authority)
+    : IpAddress(bytesFromAuthority(authority), typeFromAuthority(authority))
+{ }
 
 /**
- * Updates the byte format of IP address and type, from the string format.
+ * Updates the state of the IP object from the value of the ipString_ field
  */
-void IpAddress::toBytes() {
-    std::array<uint8_t, IpAddress::IpV6AddressBytes> bytes = {0};
-    auto length = 0;
-    if (ipString_.empty()) {
-        type_ = AddressType::Local;
-    } else if (1 == inet_pton(AF_INET, ipString_.c_str(), &bytes)) {
-        type_ = AddressType::IpV4;
-        length = IpAddress::IpV4AddressBytes;
-    } else if (1 == inet_pton(AF_INET6, ipString_.c_str(), &bytes)) {
-        type_ = AddressType::IpV6;
-        length = IpAddress::IpV6AddressBytes;
-    } else {
-        type_ = AddressType::Invalid;
-    }
+void uri::IpAddress::fromString() {
+    if (std::vector<uint8_t> bytes(IpV6AddressBytes, 0);
```

Comment:
can you put each of the cases in the if , in separate brackets?

Reply:
These if statements are in the form `if ([init-statement]; [condition])`. There is only one conditional statement within the parentheses, and typically this wouldn't require extra brackets. Are you asking me to change this to `if (([init-statement]); ([condition]))`?